Remove AllowPlacementOnResources in Building and use TerrainTypes only

michaeldgg2 opened the pull request:
This PR removes `AllowPlacementOnResources` from `BuildingInfo` and uses `BuildingInfo.TerrainTypes` only for resource-related checks. Contains an update rule, which adds all resource types defined on `ResourceLayer` to `TerrainTypes` on `BuildingInfo`, if `AllowPlacementOnResources` is true.

The last commit is just for testing the update rule, do not merge/rebase it to bleed.

pchote commented:
I went to great pains to remove all the weird hardcoded logic booleans in #19090... it looks like I missed one 🤦.

Can we please instead remove `BuildingInfo.AllowPlacementOnResources` and instead leave it to `BuildingInfo.TerrainTypes` to determine whether a building can be placed on a resource (and implement new logic for placing resources under buildings in the editor if their TerrainTypes allow it).

michaeldgg2 committed:
ResourceLayer: use BuildingInfo.TerrainTypes to check, if a resource can be placed underneath the building

michaeldgg2 committed:
Remove BuildingInfo.AllowPlacementOnResources and use TerrainTypes to check whether a building can be placed on particular cell

michaeldgg2 committed:
PBOG: remove the resource check and let IsCellBuildable() extension method handle it

michaeldgg2 commented:
I've removed `BuildingInfo.AllowPlacementOnResources` and replaced it with checks for `BuildingInfo.TerrainTypes`. Here's a test case, where I allowed Refinery to be placed on Ore and Gems:

![openra_proc_buildable_on_ore_gems](https://github.com/user-attachments/assets/554f0740-682e-453b-88a4-6a80ae718845)

PunkPun commented on OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:
```diff
@@ -23,6 +23,15 @@ public static bool IsCellBuildable(this World world, CPos cell, ActorInfo ai, Bu
+				{
+					var terrainType = world.Map.GetTerrainInfo(cell);
+					if (resourceLayer.GetResource(cell).Type != null && !bi.TerrainTypes.Contains(terrainType.Type))
+						return false;
```
isn't this check already at the bottom of the function?

PunkPun commented on OpenRA.Mods.Common/Traits/Buildings/Building.cs:
```diff
@@ -45,8 +45,6 @@ public class BuildingInfo : TraitInfo, IOccupySpaceInfo, IPlaceBuildingDecoratio
 
 		public readonly bool AllowInvalidPlacement = false;
 
-		public readonly bool AllowPlacementOnResources = false;
```
this needs an update rule

michaeldgg2 replied:
Added update rule, which adds all resource types defined on `ResourceLayer` to `TerrainTypes` on `BuildingInfo`, if `AllowPlacementOnResources` is true.

PunkPun reviewed (commented):
LGTM

PunkPun commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/bf6ee41cca6bd4cb31d6f933228be782ca34d10d...444b2fe6bf649cc54a29499f5c423414a8e3056e
